doc/release/upcoming_changes/18116.future.rst

```diff
@@ -0,0 +1,18 @@
+Promotion of strings with numbers and bools is deprecated
+---------------------------------------------------------
+Any promotion of numbers and strings is deprecated and will
+give a ``FutureWarning`` the main affected functionalities
+are:
+
+* `numpy.promote_types` and `numpy.result_type` which will raise
+  an error in this case in the future.
+* `numpy.concatenate` will raise an error when concatenating a string
+  and numeric array. You can use ``dtype="S"`` to explicitly request
+  a string result.
+* `numpy.array` and related functions will start returning ``object``
+  arrays because these functions use ``object`` as a fallback when
+  no common dtype can be found. (In this case setting the
+  ``FutureWarning`` to be raised will unfortunately lead to the new
+  behaviour)
+
+Comparisons, universal functions, and casting are not affected by this.
```

Comment:
Could you give an example of what will trigger the deprecation warning and a possible mitigation, for people who are not savvy in the ways of promotion but want to migrate their code?

Reply:
Well, added the examples from the warning message, is that what you thought of?

Reply:
Perfect.